File: SICO/SICO/SiCo.lgla2/Mantenimientos.vb
Imports SiCo.lgla
Public MustInherit Class Mantenimientos
    Inherits SiCo.lgla.Entidad

#Region "Declaraciones"
    Private _idEntidades As Integer
#End Region

#Region "Construtor"
    Public Sub New()
        MyBase.New()
        Me.ColeccionParametrosBusqueda.Add(New Parametro("identidades", Nothing))
        Me.ColeccionParametrosMantenimiento.Add(New Parametro("identidades", Nothing))

        Me.ColeccionParametrosBusqueda.Add(New Parametro("entidadnombre", Nothing))
        Me.ColeccionParametrosBusqueda.Add(New Parametro("espersonnatural", Nothing))
    End Sub
#End Region

#Region "Propiedaes"

    Public Property idEntidades() As Integer
        Get
            Return _idEntidades
        End Get
        Set(ByVal value As Integer)
            _idEntidades = value
        End Set
    End Property

    Public ReadOnly Property PersonaNatural() As PersonaNatural
        Get
            Return CrearPersonaNatural()
        End Get
    End Property

    Public ReadOnly Property PersonaJuridica() As PersonaJuridica
        Get
            Return CrearPersonaJuridica()
        End Get
    End Property


#End Region

#Region "Metodos"

    Public Overrides Sub Guardar()
        Me.ValorParametrosMantenimiento("identidades", Me.idEntidades)
        MyBase.Guardar()
    End Sub

    Public Overloads Sub Buscar(ByVal identidades As Integer)
        Me.NullParametrosBusqueda()
        Me.ValorParametrosBusqueda("identidades", identidades)
        Me.LlenadoTabla(ColeccionParametrosBusqueda)
    End Sub

    Public Overloads Sub buscar(ByVal EntidadNombre As String, ByVal EsPersonaNatural As Boolean)
        Me.NullParametrosBusqueda()
        Me.ValorParametrosBusqueda("entidadnombre", EntidadNombre)
        Me.ValorParametrosBusqueda("espersonnatural", EsPersonaNatural)
        Me.LlenadoTabla(Me.ColeccionParametrosBusqueda)
    End Sub

    Protected Overrides Sub CargadoPropiedades(ByVal Indice As Integer)
        Me.idEntidades = Me.Registro(Indice, "identidades")
        MyBase.CargadoPropiedades(Indice)
    End Sub

    Private Function CrearPersonaNatural() As PersonaNatural
        Dim i As Integer = PrimerRegistro("identidades")
        If Not i = Nothing Then
            Return New PersonaNatural(i, PrimerRegistro("entidadnombre"), New TipoIdentidad(PrimerRegistro("tipoidentificacion").ToString()), PrimerRegistro("identificacion").ToString, PrimerRegistro("correo"), PrimerRegistro("direccion"), PrimerRegistro("rtn"), PrimerRegistro("telefono"))
        End If
        Return Nothing
    End Function

    Private Function CrearPersonaJuridica() As PersonaJuridica

        Dim i As Integer = PrimerRegistro("identidades")
        If Not i = Nothing Then
            Return New PersonaJuridica(i, PrimerRegistro("entidadnombre"), PrimerRegistro("correo"), PrimerRegistro("direccion"), PrimerRegistro("rtn"), PrimerRegistro("telefono"))
        End If
        Return Nothing
    End Function

#End Region

End Class

File: SICO/SICO/SiCo.lgla/Mantenimientos.cs
using System;

namespace SiCo.lgla
{
    [Serializable]
    public abstract class Mantenimientos : Entidad
    {
        #region Declaraciones

        private int _Estado;
        private PersonaJuridica _PersonaJuridica = new PersonaJuridica();
        private PersonaNatural _PersonaNatural = new PersonaNatural();
        private long? _idEntidades;

        #endregion

        #region Constructor

        public Mantenimientos()
        {
            ColeccionParametrosBusqueda.Add(new Parametro("identidades", null));
            ColeccionParametrosBusqueda.Add(new Parametro("entidadnombre", null));
            ColeccionParametrosBusqueda.Add(new Parametro("espersonanatural", null));
            ColeccionParametrosBusqueda.Add(new Parametro("usuario", null));
            ColeccionParametrosBusqueda.Add(new Parametro("contrasena", null));
            ColeccionParametrosBusqueda.Add(new Parametro("estado", null));
            ColeccionParametrosBusqueda.Add(new Parametro("idrol", null));

            ColeccionParametrosMantenimiento.Add(new Parametro("identidades", null));
            ColeccionParametrosMantenimiento.Add(new Parametro("estado", null));

            ComandoSelect = "MantenimientosComplejos_Buscar";
        }

        public Mantenimientos(long? id, long? idEntidades, int estado)
        {
            ColeccionParametrosBusqueda.Add(new Parametro("identidades", null));
            ColeccionParametrosBusqueda.Add(new Parametro("entidadnombre", null));
            ColeccionParametrosBusqueda.Add(new Parametro("espersonanatural", null));
            ColeccionParametrosBusqueda.Add(new Parametro("usuario", null));
            ColeccionParametrosBusqueda.Add(new Parametro("contrasena", null));
            ColeccionParametrosBusqueda.Add(new Parametro("estado", null));
            ColeccionParametrosBusqueda.Add(new Parametro("idrol", null));

            ColeccionParametrosMantenimiento.Add(new Parametro("identidades", null));
            ColeccionParametrosMantenimiento.Add(new Parametro("estado", null));

            ComandoSelect = "MantenimientosComplejos_Buscar";
            _Id = id;
            _idEntidades = idEntidades;
            Estado = estado;
        }

        #endregion

        #region Propiedades

        public long? idEntidades
        {
            get { return _idEntidades; }
            set { _idEntidades = value; }
        }

        public string NombreMantenimiento
        {
            get
            {
                if (PersonaJuridica != null)
                {
                    return PersonaJuridica.RazonSocial;
                }
                if (PersonaNatural != null)
                {
                    return PersonaNatural.NombreCompletoMostrar;
                }
                return string.Empty;
            }
        }

        public PersonaNatural PersonaNatural
        {
            get
            {
                if (_PersonaNatural != null)
                {
                    if (_PersonaNatural.Id > 0)
                        return _PersonaNatural;
                }
                _PersonaNatural = CrearPersonaNatural(0);
                return _PersonaNatural;
            }

            set { _PersonaNatural = value; }
        }

        public PersonaJuridica PersonaJuridica
        {
            get
            {
                if (_PersonaJuridica != null)
                {
                    if (_PersonaJuridica.Id > 0)
                        return _PersonaJuridica;
                }
                _PersonaJuridica = CrearPersonaJuridica(0);
                return _PersonaJuridica;
            }
            set { _PersonaJuridica = value; }
        }

        public string TablaBusqueda { get; set; }

        public int Estado
        {
            get { return _Estado; }
            set { _Estado = value; }
        }

        #endregion

        #region Metodos

        public override void Guardar()
        {
            ValorParametrosMantenimiento("estado", Estado);
            ValorParametrosMantenimiento("identidades", idEntidades);
            base.Guardar();
        }

        public void Buscar(int identidades)
        {
            NullParametrosBusqueda();
            ValorParametrosBusqueda("identidades", identidades);
            LlenadoTabla(ColeccionParametrosBusqueda);
        }

        public void Buscar(string EntidadNombre)
        {
            NullParametrosBusqueda();
            ValorParametrosBusqueda("entidadnombre", EntidadNombre);
            LlenadoTabla(ColeccionParametrosBusqueda);
        }

        protected override void CargadoPropiedades(int Indice)
        {
            Estado = Convert.ToInt32(Registro(Indice, "estado"));
            idEntidades = Convert.ToInt64((Registro(Indice, "identidades")));
            _PersonaJuridica = CrearPersonaJuridica(Indice);
            _PersonaNatural = CrearPersonaNatural(Indice);

            base.CargadoPropiedades(Indice);
        }

        protected PersonaNatural CrearPersonaNatural(int Indice)
        {
            if (TotalRegistros > 0)
            {
                var p = new PersonaNatural();
                if (!(idEntidades == null) && Convert.ToInt32(Registro(Indice, "espersonanatural")) == 1)
                {
                    p = new PersonaNatural(idEntidades, (string) Registro(Indice, "entidadnombre"),
                                           new TipoIdentidad((string) Registro(Indice, "tipoidentidad")),
                                           (string) Registro(Indice, "identificacion"),
                                           (string) Registro(Indice, "correo"), (string) Registro(Indice, "direccion"),
                                           (string) Registro(Indice, "rtn"), (int?) Registro(Indice, "telefono"),
                                           (int?) Registro(Indice, "telefono2"));
                }
                else
                    p = null;


                return p;
            }
            return null;
        }

        protected PersonaJuridica CrearPersonaJuridica(int Indice)
        {
            if (TotalRegistros > 0)
            {
                var p = new PersonaJuridica();
                if (!(idEntidades == null) && Convert.ToInt32(Registro(Indice, "espersonanatural")) == 0)
                {
                    p = new PersonaJuridica(idEntidades, (string) Registro(Indice, "entidadnombre"),
                                            (string) Registro(Indice, "correo"), (string) Registro(Indice, "direccion"),
                                            (string) Registro(Indice, "rtn"), (int?) Registro(Indice, "telefono"),
                                            (int?) Registro(Indice, "telefono2"));
                }
                else
                    p = null;

                return p;
            }
            return null;
        }

        #endregion
    }
}